banyand/measure/measure_topn.go

```diff
@@ -0,0 +1,681 @@
+	// TODO: this can be optimized if the bus Listener can be synchronously finished,
+	sync.RWMutex
+	l            *logger.Logger
+	m            *measure
+	topNSchemas  []*databasev1.TopNAggregation
+	processorMap map[*commonv1.Metadata][]*topNStreamingProcessor
+}
+
+func (manager *topNProcessorManager) Close() error {
+	manager.Lock()
+	defer manager.Unlock()
+	var err error
+	for _, processorList := range manager.processorMap {
+		for _, processor := range processorList {
+			err = multierr.Append(err, processor.Close())
+		}
+	}
+	return err
+}
+
+func (manager *topNProcessorManager) onMeasureWrite(request *measurev1.WriteRequest) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	for _, processorList := range manager.processorMap {
+		for _, processor := range processorList {
+			processor.src <- flow.NewStreamRecordWithTimestampPb(request.GetDataPoint(), request.GetDataPoint().GetTimestamp())
+		}
+	}
+
+	return nil
+}
+
+func (manager *topNProcessorManager) start() error {
+	interval := manager.m.interval
+	for _, topNSchema := range manager.topNSchemas {
+		sortDirections := make([]modelv1.Sort, 0, 2)
+		if topNSchema.GetFieldValueSort() == modelv1.Sort_SORT_UNSPECIFIED {
+			sortDirections = append(sortDirections, modelv1.Sort_SORT_ASC, modelv1.Sort_SORT_DESC)
+		} else {
+			sortDirections = append(sortDirections, topNSchema.GetFieldValueSort())
+		}
+
+		processorList := make([]*topNStreamingProcessor, len(sortDirections))
+		for i, sortDirection := range sortDirections {
+			srcCh := make(chan interface{})
+			src, _ := sources.NewChannel(srcCh)
+			streamingFlow := streaming.New(src)
+
+			filters, buildErr := manager.buildFilter(topNSchema.GetCriteria())
+			if buildErr != nil {
+				return buildErr
+			}
+			streamingFlow = streamingFlow.Filter(filters)
+
+			mapper, innerErr := manager.buildMapper(topNSchema.GetFieldName(), topNSchema.GetGroupByTagNames()...)
+			if innerErr != nil {
+				return innerErr
+			}
+			streamingFlow = streamingFlow.Map(mapper)
+
+			processor := &topNStreamingProcessor{
+				l:                manager.l,
+				shardNum:         manager.m.shardNum,
+				interval:         interval,
+				topNSchema:       topNSchema,
+				sortDirection:    sortDirection,
+				databaseSupplier: manager.m.databaseSupplier,
+				src:              srcCh,
+				in:               make(chan flow.StreamRecord),
+				stopCh:           make(chan struct{}),
+				streamingFlow:    streamingFlow,
+			}
+			processorList[i] = processor.start()
+		}
+
+		manager.processorMap[topNSchema.GetSourceMeasure()] = processorList
+	}
+
+	return nil
+}
+
+func (manager *topNProcessorManager) buildFilter(criteria *modelv1.Criteria) (flow.UnaryFunc[bool], error) {
+	// if criteria is nil, we handle all incoming elements
+	if criteria == nil {
+		return func(_ context.Context, dataPoint any) bool {
+			return true
+		}, nil
+	}
+
+	f, err := manager.buildFilterForCriteria(criteria)
```

Comment:
`pkg/query/logical/tag_filter` provided a enhanced filter. Please use it simplify the manager.

Reply:
Yep. It's not easy, but we have to. The match operation is complex, we should not provide two variants here. The code will become more complicated to maintain from time to time. Furthermore, the topn filter supports a subset of criteria. We must enhance it in the future if we still select this path.

A convenient way is to transform `[]*model_v1.TagFamilyForWrite` to `[]*model_v1.TagFamily` by appending `name` to them.